### PR TITLE
Support for behavior.on.error, DLQ and Retries

### PR DESCRIPTION
**Pull request by fahadhasher**
PR adds support for the following features:

1. **Error-tolerance** : User-defined behavior of the Connector while handling exceptions.
2. **DLQ** : Support for writing failed records to a user-defined Kafka topic.
3. **Retries** : Retrying(with constant/exponential backOff) in case of transient exceptions being raised while ingesting records into KustoDB.

Configs added:
```
behavior.on.error=FAIL
dlq.bootstrap.servers=localhost:9092
dlq.topic.name=test-topic-error
errors.retry.max.time.ms=60000
errors.retry.backoff.time.ms=5000
```

Removed username/password authentication.
Added connector level temporary directory creation logic for storing tmp files.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -71,7 +71,7 @@
         <kafka.version>1.0.0</kafka.version>
         <json.version>20090211</json.version>
         <commonio.version>2.6</commonio.version>
-        <kusto-sdk.version>2.1.0</kusto-sdk.version>
+        <kusto-sdk.version>1.4.2</kusto-sdk.version>
```
Comment (vladikbr) [inline pom.xml:74]: are we downgrading the SDK version?

Comment (fahadhasher) [reply pom.xml:74]: We were facing issues while performing table access vaildation with the new SDK.
Currently discussing this with Ohad.
We'll be using the latest version of the SDK.
Also, we are verifying these features, will let you know when it's completed and ready to be reviewed.

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 2]
```diff
@@ -15,134 +20,258 @@
...
 
-            return IngestClientFactory.createClient(ConnectionStringBuilder.createWithAadUserCredentials(
-                    config.getKustoUrl(),
-                    config.getAuthUsername(),
-                    config.getAuthPassword()
-            ));
-        }
 
-        throw new ConfigException("Kusto authentication method must be provided.");
+            throw new ConnectException("Failed to initialize KustoIngestClient, please " +
```
Comment (Sumanthonnavar) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:92]: This exception supposed to be ConfigException?

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 3]
```diff
@@ -15,134 +20,258 @@
...
+                if (Strings.isNullOrEmpty(config.getAuthPassword())) {
+                    throw new ConfigException("Kusto authentication missing Password.");
+                }
+                return ClientFactory.createClient(ConnectionStringBuilder.createWithAadUserCredentials(
+                        engineClientURL,
+                        config.getAuthUsername(),
+                        config.getAuthPassword()
+                ));
+            }
+            throw new ConnectException("Failed to initialize KustoEngineClient, please " +
```
Comment (Sumanthonnavar) [inline src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:128]: This exception supposed to be ConfigException?

File: README.md [hunk 4]
```diff
@@ -50,21 +50,40 @@ Go to `http://localhost:3030/kafka-connect-ui/#/cluster/fast-data-dev/` and usin
...
 key.converter=org.apache.kafka.connect.storage.StringConverter 
 value.converter=org.apache.kafka.connect.storage.StringConverter 
+
 tasks.max=1 
-topics=testing1 
-kusto.tables.topics_mapping=[{'topic': 'testing1','db': 'daniel', 'table': 'KafkaTest','format': 'json', 'mapping':'JsonMapping'},{'topic': 'testing2','db': 'daniel', 'table': 'KafkaTest','format': 'csv', 'mapping':'CsvMapping', 'eventDataCompression':'gz'},] 
-kusto.auth.authority=XXX 
+topics=test-topic 
+
+kusto.tables.topics.mapping=[{'topic': 'testing1','db': 'daniel', 'table': 'KafkaTest','format': 'json', 'mapping':'JsonMapping'},{'topic': 'testing2','db': 'daniel', 'table': 'KafkaTest','format': 'csv', 'mapping':'CsvMapping', 'eventDataCompression':'gz'},]
```
Comment (anagha-microsoft) [inline README.md:63]: We have a topic called "test-topic" but referencing "testing1" and "testing-2"  in the mapping.  Can we correct this please?

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/TopicPartitionWriter.java [hunk 5]
```diff
@@ -16,72 +24,134 @@
...
+       * avoid consumer leaving the group while the Connector is retrying.
+       */
+      for (int retryAttempts = 0; true; retryAttempts++) {
+          try {
+              client.ingestFromFile(fileSourceInfo, ingestionProps);
+              log.info(String.format("Kusto ingestion: file (%s) of size (%s) at current offset (%s)", fileDescriptor.path, fileDescriptor.rawBytes, currentOffset));
+              this.lastCommittedOffset = currentOffset;
+              return;
+          } catch (IngestionServiceException | IngestionClientException exception) {
+              // TODO : improve handling of specific transient exceptions once the client supports them.
```
[the remark this answers is not in the document]
Comment (vladikbr) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/TopicPartitionWriter.java:103]: I agree.

File: src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java [hunk 6]
```diff
@@ -173,42 +274,44 @@ public void close(Collection<TopicPartition> partitions) {
...
-            maxFileSize = config.getFlushSizeBytes();
-            flushInterval = config.getFlushInterval();
-            log.info(String.format("Kafka Kusto Sink started. target cluster: (%s), source topics: (%s)", url, topicsToIngestionProps.keySet().toString()));
+    public void start(Map<String, String> props) {
+        
+        config = new KustoSinkConfig(props);
+        String url = config.getKustoUrl();
+      
+        // TODO : revisit the implementation
+        // validateTableMappings(config);
```
[the remark this answers is not in the document]
Comment (fahadhasher) [reply src/main/java/com/microsoft/azure/kusto/kafka/connect/sink/KustoSinkTask.java:290]: we did test it and it's working as expected. Uncommenting it.

**Pull request merged**